x-pack/test/functional/services/ml/data_frame_analytics_edit.ts

```diff
@@ -7,17 +7,218 @@
+          'mlJobCustomUrlTimeRangeInput',
+          customUrl.timeRange
+        );
+      }
+      if (customUrl.timeRange === TIME_RANGE_TYPE.INTERVAL) {
+        await customUrls.setCustomUrlTimeRangeInterval(customUrl.timeRangeInterval!);
+      }
+    },
+
+    async fillInDashboardUrlForm(customUrl: DashboardUrlConfig) {
+      await this.clickOpenCustomUrlEditor();
+      await customUrls.setCustomUrlLabel(customUrl.label);
+      await mlCommonUI.selectRadioGroupValue(
+        'mlJobCustomUrlLinkToTypeInput',
+        URL_TYPE.KIBANA_DASHBOARD
+      );
+      await mlCommonUI.selectSelectValueByVisibleText(
+        'mlJobCustomUrlDashboardNameInput',
+        customUrl.dashboardName
+      );
+
+      await customUrls.setCustomUrlQueryEntityFieldNames(customUrl.queryEntityFieldNames);
+      if (customUrl.timeRange === TIME_RANGE_TYPE.INTERVAL) {
+        await this.clickIntervalTimeRangeSwitch();
+        await customUrls.setCustomUrlTimeRangeInterval(customUrl.timeRangeInterval!);
+      }
+    },
+
+    async fillInOtherUrlForm(customUrl: OtherUrlConfig) {
+      await this.clickOpenCustomUrlEditor();
+      await customUrls.setCustomUrlLabel(customUrl.label);
+      await mlCommonUI.selectRadioGroupValue('mlJobCustomUrlLinkToTypeInput', URL_TYPE.OTHER);
+      await customUrls.setCustomUrlOtherTypeUrl(customUrl.url);
+    },
+
+    async addDashboardCustomUrl(
+      jobId: string,
+      customUrl: DashboardUrlConfig,
+      expectedResult: { index: number; url: string }
+    ) {
+      await this.openEditCustomUrlsForJobTab(jobId);
+
+      await retry.tryForTime(30 * 1000, async () => {
+        await this.fillInDashboardUrlForm(customUrl);
+      });
+      await this.saveCustomUrl(customUrl.label, expectedResult.index, expectedResult.url);
+      await this.updateAnalyticsJob();
+    },
+
+    async addDiscoverCustomUrl(jobId: string, customUrl: DiscoverUrlConfig) {
+      await this.openEditCustomUrlsForJobTab(jobId);
+
+      await retry.tryForTime(30 * 1000, async () => {
+        const existingCustomUrlCount = await this.getExistingCustomUrlCount();
+
+        await this.fillInDiscoverUrlForm(customUrl);
+        await this.saveCustomUrl(customUrl.label, existingCustomUrlCount);
+      });
+
+      await this.updateAnalyticsJob();
+    },
+
+    async addOtherTypeCustomUrl(jobId: string, customUrl: OtherUrlConfig) {
+      await this.openEditCustomUrlsForJobTab(jobId);
+
+      await retry.tryForTime(30 * 1000, async () => {
+        const existingCustomUrlCount = await this.getExistingCustomUrlCount();
+
+        await this.fillInOtherUrlForm(customUrl);
+        await this.saveCustomUrl(customUrl.label, existingCustomUrlCount);
+      });
+
+      await this.updateAnalyticsJob();
+    },
+
+    async testOtherTypeCustomUrlAction(indexInList: number, expectedUrl: string) {
+      await customUrls.assertCustomUrlUrlValue(indexInList, expectedUrl);
+      await this.closeEditJobFlyout();
```

Comment:
Similar to the addUrl functions, we can potentially do the following within this function?

- openEditCustomUrlsForJobTab 
- assertCustomUrlUrlValue
- closeEditJobFlyout